pkg/suggestion/v1beta1/pbt/service.py

```diff
@@ -0,0 +1,392 @@
+                request_count = len(self.job_queue)
+            elif len(self.job_queue.running) > 0:
+                logger.info(
+                    "Waiting for trials to complete before spawning next generation: {}".format(
+                        list(self.job_queue.running.keys())
+                    )
+                )
+                self.job_queue.verify_running(request.trials)
+                return api_pb2.GetSuggestionsReply(parameter_assignments=[])
+            else:
+                logger.info("Spawning next generation...")
+                self.job_queue.generate()
+
+        jobs = [self.job_queue.get() for _ in range(request_count)]
+        parameter_assignments = Assignment.generate([j[0] for j in jobs])
+        annotations = Annotations.generate([j[1] for j in jobs])
+        logger.info("Transmitting suggestion...")
+
+        return api_pb2.GetSuggestionsReply(
+            parameter_assignments=parameter_assignments, annotations=annotations
+        )
+
+    def _set_validate_context_error(self, context, error_message):
+        context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+        context.set_details(error_message)
+        logger.info(error_message)
+        return api_pb2.ValidateAlgorithmSettingsReply()
+
+
+class HyperParameterSampler(HyperParameter):
+    def __init__(self, *args, **kwargs):
+        super(HyperParameterSampler, self).__init__(*args, **kwargs)
+        if self.type in (constant.INTEGER, constant.DOUBLE):
+            self.sample_list = np.arange(
+                float(self.min),
+                float(self.max) + float(self.step) / 2,
+                float(self.step),
+            ).astype(int if self.type == constant.INTEGER else float)
+        else:
+            self.sample_list = self.list
+
+    def sample(self):
+        return np.random.choice(self.sample_list, 1)[0]
+
+    def perturb(self, value):
+        if self.type == constant.INTEGER:
+            new_value = int(int(value) * np.random.choice([0.8, 1.2], 1)[0])
+            return max(float(self.min), min(float(self.max), new_value))
+        elif self.type == constant.DOUBLE:
+            new_value = float(value) * np.random.choice([0.8, 1.2], 1)[0]
+            return max(float(self.min), min(float(self.max), new_value))
+        else:
+            sample_index = (
+                self.sample_list.index(value) + np.random.choice([-1, 1], 1)[0]
+            )
+            if sample_index >= len(self.sample_list):
+                return self.sample_list[0]
+            else:
+                return self.sample_list[sample_index]
+
+    @staticmethod
+    def from_hyperparameter(hp: HyperParameter):
+        return HyperParameterSampler(hp.name, hp.type, hp.min, hp.max, hp.list, hp.step)
+
+
+class PbtJob(object):
+    def __init__(self, assignment_list, generation=0, parent=None, uid=None):
+        if uid is None:
+            self.uid = str(uuid.uuid4())
+        else:
+            self.uid = uid
```

Comment:
Are we using UUID here to define Trial name from the Suggestion ?
For some use-cases Suggestion has to know the generated Trial name, I created the issue before: https://github.com/kubeflow/katib/issues/610.
Do we want to move forward with this feature or we use UUID for something else @a9p ?
cc @gaocegege @johnugeorge

Reply:
Correct, the uid is used to track the Trial from spawn to result. `I think, we should provide functionality, when user can specify special name to create Trial.` in that referenced issue would be necessary for PBT to be able to use that mechanism instead of the uid annotation.

Reply:
Let's see what others think about this feature.
cc @tenzen-y

Reply:
I am not sure of use-cases that would specify a special name for Trial.
So, I think it would be great to add uid annotation.
WDYT @andreyvelich @a9p ?

Reply:
For example, since BO algorithm doesn't generate Trial name, we have to search for HyperParameters based on Trial assignments: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/skopt/base_service.py#L84-L98.

If Suggestion can generate Trial names, we can store these names in `recorded_trials_names` once SKOpt service [suggests new HPs](https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/skopt/base_service.py#L121).
Once `GetSuggestions` called again, we can easily update loss value in `recorded_trials_names` based on Trial name.

Reply:
Moved to specifying trial name in the Reply rather than the Annotation per your linked issue.

Reply:
Thanks for clarifying, @andreyvelich!
That feature will help us.